### PR TITLE
Gif support, no audio codec support, quality changes

### PR DESCRIPTION
**Pull request by ghost**
Added gif & no audio video support
Increased bitrate to 512k for video
Increased bitrate to 320k for audio
Set higher quality as default
Fix wrong width crash

### CODE REVIEW

File: dpgv4.py [hunk 1]
```diff
@@ -321,8 +321,10 @@ def prepare_video_conversion_command(
         "-f", "data",
         "-map", "0:v:0",
         "-r", f"{framerate:g}",
-        "-sws_flags", "lanczos",
-        "-s", f"{width}x{height}",
+        #"-sws_flags", "lanczos",
+        #"-s", f"{width}x{height}",
```
Comment (pawel-slowik) [inline dpgv4.py:325]: Please remove these two lines instead of commenting them out. Nothing is gained by keeping them.

File: dpgv4.py [hunk 2]
```diff
@@ -321,8 +321,10 @@ def prepare_video_conversion_command(
...
+        #Prevent wrong width crash
+        "-vf", "scale=min(256\\,192*a):-1,pad=256:192:abs(ow-iw)/2:abs(oh-ih)/2",
```
Comment (pawel-slowik) [inline dpgv4.py:327]: Please use [f-strings](https://docs.python.org/3/tutorial/inputoutput.html#formatted-string-literals) and replace 256 and 192 with `SCREEN_WIDTH` and `SCREEN_HEIGHT` respectively.

Since the `width` and `heigt` variables aren't going to be used anymore, please remove them along with the `calculate_dimensions` function and its test.

An alternative approach for this could be to keep the `calculate_dimensions` function and expand it to calculate not only width and height, but also vertical and horizontal padding. For example, return a named tuple with `width`, `height`, `horizontal_padding` and `vertical_padding` fields, use `width` and `height` for the `scale` filter and `vertical_padding` and `horizontal_padding` for the `pad` filter. This would have the advantage of being more testable.

File: dpgv4.py [hunk 3]
```diff
@@ -343,7 +345,7 @@ def prepare_audio_conversion_command(input_file: str, aid: Optional[int]) -> Seq
         "-f", "data",
         "-map", f"0:a:{0 if aid is None else aid}",
         "-c:a", "mp2",
-        "-b:a", "128k",
+        "-b:a", "320k",
```
Comment (pawel-slowik) [inline dpgv4.py:348]: This change is not acceptable for me because of the output file size increase it entails. Please make it dependent on high quality mode (make the `prepare_audio_conversion_command` function accept a `quality` parameter and set the bitrate accordingly).

File: dpgv4.py [hunk 4]
```diff
@@ -364,7 +366,7 @@ def video_quality_options() -> Mapping[int, Iterable[str]]:
         "-trellis", "1",
         "-mpv_flags", "+cbp_rd",
         "-mpv_flags", "+mv0",
-        "-b:v", "256k",
+        "-b:v", "512k",
```
Comment (pawel-slowik) [inline dpgv4.py:369]: This change is not acceptable for me because of the output file size increase it entails. Please make it dependent on high quality mode.

File: dpgv4.py [hunk 5]
```diff
@@ -551,11 +553,24 @@ def convert_file(input_file: str, output_file: str, options: Any) -> None:
...
+    try:
+        a_cmd = prepare_audio_conversion_command(input_file, options.aid)
+        a_tmp_file = TemporaryFile()
+        encode_stream("audio", a_cmd, a_tmp_file)
+    except ExternalCommandFailedError:
+        a_cmd = prepare_audio_conversion_command(input_file, options.aid)
+        a_tmp_file = TemporaryFile()
+        for i in list(reversed(['-t',str(count_video_frames(v_tmp_file) / options.framerate),'-f','lavfi','-i','anullsrc'])):
+            a_cmd.insert(1,i)
+        encode_stream("audio", a_cmd, a_tmp_file)
```
Comment (pawel-slowik) [inline dpgv4.py:566]: This is not the correct way to handle missing audio, because it would run on all audio encoding errors instead of only when the audio stream is missing. This would hide other, unrelated errors and make them harder to debug.

We need to check whether the input file has an audio stream *before* attempting to encode it, i.e. in the `prepare_audio_conversion_command` function. This could be done with a `-show_streams` call and then a search for specific codec type. The function should then return either a command for encoding the existing audio stream or a command that creates a null stream.

The `list(reversed` loop is hard to read, please use list slicing instead. If you need to insert a list into another list starting at position 1: 
`a_cmd = a_cmd[:1] + [whatever, you, need, to, insert] + a_cmd[1:]` 
But this is likely unnecessary, just create the null audio encoding command from scratch in `prepare_audio_conversion_command` instead of modifying an inadequate command.

File: dpgv4.py [hunk 6]
```diff
@@ -551,11 +553,24 @@ def convert_file(input_file: str, output_file: str, options: Any) -> None:
...
     gop = create_gop(v_tmp_file)
-    thumbnail = create_thumbnail(create_screenshot(v_tmp_file, int(get_duration(input_file) / 10)))
+    #Catch bad thumbnails
+    try:
+        thumbnail = create_thumbnail(create_screenshot(v_tmp_file, int(get_duration(input_file) / 10)))
+    except:
+        print("Bad Thumbnail")
+        exit()
```
Comment (pawel-slowik) [inline dpgv4.py:573]: 1) This would not only catch "bad thumbnails" (whatever that means), but also errors originating from the `create_screenshot` and `get_duration` functions. For example, if the ffmpeg duration command crashed for some reason, we'd get a misleading "Bad Thumbnail" message. This would only make it harder to find the real reason for the failure.

2) Why catch the error at all if we're not doing anything about it? As I understand it, we'd get a short "Bad Thumbnail" message instead of an exception traceback. This is useless, because this short message doesn't help with debugging, while the exception does.

3) The message would be printed on the [stdout](https://en.wikipedia.org/wiki/Standard\_streams#Standard\_output_(stdout)) instead of stderr, which is bad practice for error messages.

4) The program's exit code in case of caught bad thumbnail would change to 0, which indicates successfull completion, while the program in fact did not successfully complete its task.

To sum up: this change is not going to be applied.

If you have a need to handle errors during thumbnail generation, please explain how would you like the error to be handled. We could use some kind of a placeholder thumbnail, but handling errors via hiding them is not acceptable.

File: dpgv4.py [hunk 7]
```diff
@@ -717,8 +732,8 @@ def main() -> None:
     )
     video_group = parser.add_argument_group("video options")
     video_group.add_argument(
-        "-q", type=int, dest="quality", choices=video_quality_options().keys(), default=1,
-        help="quality setting (default: 1)"
+        "-q", type=int, dest="quality", choices=video_quality_options().keys(), default=2,
+        help="quality setting (default: 2)"
```
Comment (pawel-slowik) [inline dpgv4.py:736]: This change is not going to happen. Lowest quality is the default, because we're encoding for a very low screen resolution, which means that any "high quality" options are unlikely to result in a perceptible improvement. This may be somewhat dependent on the source file, but if that's the case, just use `-q 2` instead of forcing it as default for all users.

**Comment by pawel-slowik** [comment]
Hi @Wisherite, thank you for your interest in dpgv4.

This PR is not going to get merged because it rolls several unrelated changes, some of which I disagree with, into one package. Also, the commits are not atomic, which makes cherry-picking impossible. Therefore, it needs to be rewritten anyway, and if you're rewriting, you could as well start from scratch instead of jumping through hoops with rebasing.

Please feel free to resubmit the changes related to:
- vertically filmed videos,
- bitrate increases,
- handling of missing audio

after updating the solutions to take the above comments into account.

This could be either three separate pull requests or a single pull request, but with commits focused on a single topic. Please don't use `Update dpgv4.py` as a commit subject.

The two other changes:
- change of the default quality setting,
- catching "bad thumbnails"

are not acceptable for reasons explained above.

If you do not wish to rework it, that's okay too. It may take me some time, but I'll get to fixing the errors eventually.
